### DISCUSSION TITLE
[FEAT] customization of backdrop

### DISCUSSION THREAD

**Discussion by felippewick** [lodev09/react-native-true-sheet#44] [Ideas]
Hey!

First of all, thank you for this awesome library! 
I was wondering if there are any plans to allow the customization of the backdrop (view behind the sheet)? Like color and opacity? 

That would be awesome, especially for smaller sheets.

Thanks a lot!

**Comment by lodev09** [comment]
Unfortunately, this isn't possible on IOS, natively. Hopefully this becomes possible on future versions but I really doubt it.

**Reply by felippewick** [reply]
That's a bummer. But thanks for your quick reply!

**Reply by hirbod** [reply]
Haha @felippewick, I am wrecking around my head here too.

**Comment by lodev09** [comment]
I did play with modifying the `UIDimmedView` but it requires lots of code/modification on how IOS handles the dimming. I'm able to change the alpha but it would require other modifications to work with the animations (e.g. dragging the sheet and animate the dimming).

Still not worth it at this time. However, I'm open to PRs/suggestions

**Reply by hirbod** [reply]
@lodev09 support for the alpha would already be huge. I added some ugly hacks with a custom view, but since I am not aware of the progress value and I have no progress while dragging, everything is just "hacky". It works fine when the close is initiated by tap, but on drag, it's delayed. See my video. 

https://github.com/lodev09/react-native-true-sheet/assets/504909/310e627c-5b15-451f-91a0-b2c278e19b44

**Reply by hirbod** [reply]
Here is a better video showcasing why I am not very happy with the result, since dragging does not emit any progress, thus I can't interpolate. If we could define the alpha (we definitely need darker values here), I'd already be super happy!

https://github.com/lodev09/react-native-true-sheet/assets/504909/025ab272-3596-41dc-8925-3cc26b5884e3

**Reply by lodev09** [reply]
You can try it in this block:
https://github.com/lodev09/react-native-true-sheet/blob/ecc83c098537c9f8227c42853a8b7c1002c8eed4/ios/TrueSheetViewController.swift#L120-L129

This override is called when you drag the sheet which is kinda weird :D

**Reply by hirbod** [reply]
If you are able to implement a progress value callback (ideally a useSharedValue) from -1 to any index and vice versa, we could just interpolate and clamp it in user land and build a custom backdrop component.

**Reply by hirbod** [reply]
I'll add 150 USD bounty on that request. Needs to work on drag for both iOS and Android, and of course when it's closing/opening.

**Reply by lodev09** [reply]
> If you are able to implement a progress value callback (ideally a useSharedValue) from -1 to any index and vice versa, we could just interpolate and clamp it in user land and build a custom backdrop component.

I like the idea of a user providing a custom backdrop component instead. I can open up a PR to expose an `onDrag` event. I believe reanimated has a feature to wrap a nativeEvent

**Reply by lodev09** [reply]
@hirbod this is available (in main branch as of this writing). You can now utilize drag events and it does support reanimated. Just need to add dim prop and utilize `useAnimatedProps` from Reanimated :)

**Reply by hirbod** [reply]
@lodev09 that is lovely. We're not using your library anymore unfortunately. Some features have been strongly requested long time ago and we went back to Gorhom.

I'll still give it a try though and see if we could use this again in our new app.

**Comment by mikhail-create** [comment]
Hey!

We faced a backdrop customization issue requested by our customer. We implemented two solutions for iOS.

The first one involves a patch for `TrueSheetViewController.swift`. In this case, we can only manipulate the opacity, but it meets our requirements. 

```swift
override func viewWillAppear(_ animated: Bool) {
    super.viewWillAppear(animated)
    UIView.animate(withDuration: 0.4) {
        self.presentingViewController?.view.alpha = 0.2
    }
    delegate?.viewControllerWillAppear()
}

override func viewWillDisappear(_ animated: Bool) {
    super.viewWillDisappear(animated)
    UIView.animate(withDuration: 0.4) {
        self.presentingViewController?.view.alpha = 1
    }
}
```

The second solution involves creating a `BottomSheetProvider` component, which includes the backdrop component with animations. In this case, we also added a new prop, `onWillPresent`, to achieve smooth backdrop animations. Feel free to ask for some code examples of the provider component.

I hope this helps someone out there or sparks new ideas for their solutions.

**Reply by lodev09** [reply]
I just merged PR #124 to support drag events and Reanimated. If I can expose an `dimAlpha` prop, could probably take advantage of `Animated.createAnimatedComponent` and pass the shared value to that prop -- I'm not 100% sure if that's how reanimated works tho.

`onWillPresent` and `onWillDismiss` is in my next to-do #132

**Reply by lodev09** [reply]
ah yeah, Reanimated does support native props. I have tried this before with the Camera
https://docs.swmansion.com/react-native-reanimated/docs/core/useAnimatedProps/

**Comment by lodev09** [comment]
Fabric #211 is now supported with all these changes applied! Also background improvements at #214 🥳
